ToolBox/TradingEconomicsDataDownloader/TradingEconomicsCalendarDownloader.cs

```diff
@@ -48,52 +55,129 @@ public TradingEconomicsCalendarDownloader(string destinationFolder)
         /// <returns>True if process all downloads successfully</returns>
         public override bool Run()
         {
+            Log.Trace("TradingEconomicsCalendarDownloader.Run(): Begin downloading calendar data");
+
             var stopwatch = Stopwatch.StartNew();
             var data = new List<TradingEconomicsCalendar>();
+            var availableFiles = Directory.GetFiles(_destinationFolder, "*.zip", SearchOption.AllDirectories)
+                .Select(
+                    x =>
+                    {
+                        try
+                        {
+                            return DateTime.ParseExact(Path.GetFileName(x).Substring(0, 8), "yyyyMMdd", CultureInfo.InvariantCulture);
+                        }
+                        catch
+                        {
+                            return DateTime.MinValue;
+                        }
+                    }
+                )
+                .Where(x => x != DateTime.MinValue)
+                .ToHashSet();
 
             var startUtc = _fromDate;
             while (startUtc < _toDate)
             {
                 try
                 {
                     var endUtc = startUtc.AddMonths(1).AddDays(-1);
+
+                    Log.Trace($"TradingEconomicsCalendarDownloader.Run(): Collecting calendar data from {startUtc:yyyy-MM-dd} to {endUtc:yyyy-MM-dd}");
+
+                    if (availableFiles.Contains(endUtc))
+                    {
+                        Log.Trace($"TradingEconomicsCalendarDownloader.Run(): Skipping data because it already exists for month: {startUtc:MMMM}");
+                        startUtc = startUtc.AddMonths(1);
+                        continue;
+                    }
+
+                    _requestGate.WaitToProceed(TimeSpan.FromSeconds(1));
+
                     var content = Get(startUtc, endUtc).Result;
                     var collection = JsonConvert.DeserializeObject<List<TradingEconomicsCalendar>>(content);
 
-                    data.AddRange(collection);
+                    // Only write data that contains the "actual" field so that we get the final
```

Comment:
Can you please clarify this comment?

What the "actual" field means and why it getting the "final piece of unchanging data" can affect the backward consistency?

Reply:
For future reference:

TE goes back and updates their dataset whenever they get data for things like Unemployment, Productivity, etc. However, they do not have any historical snapshots for how data was before they went back and updated the data. Because of this, we decided that the best way forward to maintain consistency with the data in live trading and backtesting was to only include published data instead of data for an event that hasn't occurred yet.